Add Custom error message support for Toml Schema validator

xlight05 opened the pull request:
## Purpose
$Subject.

Fixes #27903 

## Approach
Introduced a seperate field for each schema, called `message`. Message can contain error messages for all the validations. Keys of the message field is equals to the validation key. See the sample below.  and check `CustomErrorTest.java` for usage. 

## Samples
```json 
{
    "$schema": "http://json-schema.org/draft-07/schema#",
    "title": "Ballerina Manifest Spec",
    "description": "Schema for Ballerina Manifest",
    "type": "object",
    "additionalProperties": false,
    "properties": {
        "package": {
            "type": "object",
            "additionalProperties": false,
            "properties": {
                "name": {
                    "type": "string",
                    "message" : {
                        "type" : "Invalid Type"
                    }
                },
                "org": {
                    "type": "string",
                    "pattern": "^(?!\\s*$).+",
                    "message" : {
                        "pattern" : "Org cant be empty"
                    }
                },
                "version": {
                    "type": "string",
                    "pattern": "^(?!\\s*$).+"
                },
                "port": {
                    "type": "integer",
                    "minimum": 0,
                    "message" : {
                        "minimum" : "Port should be positive int"
                    }
                },
                "capacity": {
                    "type": "integer",
                    "maximum": 5,
                    "message" : {
                        "maximum" : "Max Capacity is 5"
                    }
                },
                "repository": {
                    "type": "string"
                },
                "keywords": {
                    "type": "array",
                    "items": {
                        "type": "string"
                    }
                },
                "exported": {
                    "type": "array",
                    "items": {
                        "type": "string"
                    }
                }
            },
            "required": ["name","org"],
            "message" : {
                "required" : "Field ${property} is required",
                "additionalProperties" : "Field ${property} is not supported"
            }
        }
    }
}

```

## Remarks
> List any other known issues, related PRs, TODO items, or any other notes related to the PR.

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [ ] Added necessary tests
  - [x] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [x] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

pramodya1994 commented on misc/toml-parser/src/main/java/io/ballerina/toml/validator/SchemaValidator.java:
```diff
@@ -55,6 +55,8 @@
  */
 public class SchemaValidator extends TomlNodeVisitor {
 
+    public static final String PROPERTY_HOLDER = "${property}";
```
Access can be private?

pramodya1994 commented on misc/toml-parser/src/main/java/io/ballerina/toml/validator/SchemaValidator.java:
```diff
@@ -66,12 +68,11 @@ public SchemaValidator(Schema schema) {
     public void visit(TomlTableNode tomlTableNode) {
         if (schema.type() != Type.OBJECT) {
             TomlDiagnostic diagnostic = getTomlDiagnostic(tomlTableNode.location(), "TVE0002", "error.invalid.type",
```
Shall we define constants for `"TVE0002"` and `"error.invalid.type"`?

xlight05 replied:
This should be addressed with templates with all the error messages. We've created an issue to keep track of this. #27041
Shall we do this change with that PR?

pramodya1994 commented on misc/toml-parser/src/main/java/io/ballerina/toml/validator/schema/Schema.java:
```diff
@@ -58,9 +65,7 @@ public static Schema from(Path jsonPath) throws IOException {
-        Schema rootSchema = gson.fromJson(reader, Schema.class);
-        rootSchema.setType(Type.OBJECT);
-        return rootSchema;
+        return (Schema) gson.fromJson(reader, AbstractSchema.class);
```
Seems like this casting is unnecessary

xlight05 replied:
Casting is needed as the static method should return a ObjectSchema(Schema) instead of AbstractSchema. 
Note in the old code we called `gson.fromJson(reader, Schema.class);` and now we call `gson.fromJson(reader, AbstractSchema.class);`

pramodya1994 commented on misc/toml-parser/src/main/java/io/ballerina/toml/validator/schema/Schema.java:
```diff
@@ -72,16 +77,30 @@ public static Schema from(Path jsonPath) throws IOException {
-        Schema rootSchema = gson.fromJson(jsonContent, Schema.class);
-        rootSchema.setType(Type.OBJECT);
-        return rootSchema;
+        return (Schema) gson.fromJson(jsonContent, AbstractSchema.class);
```
Seems like this casting is unnecessary

xlight05 replied:
Casting is needed as the static method should return a ObjectSchema(Schema) instead of AbstractSchema. 
Note in the old code we called `gson.fromJson(reader, Schema.class);` and now we call `gson.fromJson(reader, AbstractSchema.class);`

pramodya1994 commented on misc/toml-parser/src/main/java/io/ballerina/toml/validator/schema/SchemaDeserializer.java:
```diff
@@ -162,6 +175,51 @@ private String parseOptionalStringFromJson(JsonObject jsonObject, String key) {
+        if (customMessageJson == null) {
+            return new HashMap<>();
+        }
+        JsonElement type = customMessageJson.get(TYPE);
+        if (!(type == null || type.isJsonNull())) {
+            String customTypeMessage = type.getAsString();
+            customMessages.put(TYPE, customTypeMessage);
+        }
+
+        JsonElement pattern = customMessageJson.get(PATTERN);
+        if (!(pattern == null || pattern.isJsonNull())) {
+            String customPatternMessage = pattern.getAsString();
+            customMessages.put(PATTERN, customPatternMessage);
+        }
+
+        JsonElement required = customMessageJson.get(REQUIRED);
+        if (!(required == null || required.isJsonNull())) {
+            String customRequiredMessage = required.getAsString();
+            customMessages.put(REQUIRED, customRequiredMessage);
+        }
+
+        JsonElement additionalProperties = customMessageJson.get(ADDITIONAL_PROPERTIES);
+        if (!(additionalProperties == null || additionalProperties.isJsonNull())) {
+            String customAdditionalPropMessage = additionalProperties.getAsString();
+            customMessages.put(ADDITIONAL_PROPERTIES, customAdditionalPropMessage);
+        }
+
+        JsonElement minimum = customMessageJson.get(MINIMUM);
+        if (!(minimum == null || minimum.isJsonNull())) {
+            String customMinMessage = minimum.getAsString();
+            customMessages.put(MINIMUM, customMinMessage);
+        }
+
+        JsonElement maximum = customMessageJson.get(MAXIMUM);
+        if (!(maximum == null || maximum.isJsonNull())) {
+            String customMaxMessage = maximum.getAsString();
+            customMessages.put(MAXIMUM, customMaxMessage);
+        }
```
Can we replace this code with follwoing method to avoid duplicates?
```java
private void addFieldToCustomMessagesMap(Map<String, String> customMessages, JsonObject customMessageJson,
            String field) {
        JsonElement type = customMessageJson.get(field);
        if (!(type == null || type.isJsonNull())) {
            String customTypeMessage = type.getAsString();
            customMessages.put(field, customTypeMessage);
        }
    }
```
Then you can invoke
```java
        addFieldToCustomMessagesMap(customMessages, customMessageJson, TYPE);
        addFieldToCustomMessagesMap(customMessages, customMessageJson, PATTERN);
        addFieldToCustomMessagesMap(customMessages, customMessageJson, REQUIRED);
        addFieldToCustomMessagesMap(customMessages, customMessageJson, ADDITIONAL_PROPERTIES);
        addFieldToCustomMessagesMap(customMessages, customMessageJson, MINIMUM);
        addFieldToCustomMessagesMap(customMessages, customMessageJson, MAXIMUM);

pramodya1994 reviewed (changes requested):
Pls check comments added

pubudu91 commented:
What would the full form of these errors look like when printed on the console?

xlight05 commented:
> Pls check comments added
Thanks. Addressed the issues, please check

> What would the full form of these errors look like when printed on the console?
When we add the following changes to C2C(Kubernetes.toml)
```toml
"repository": {
    "type": "string",
    "pattern": "^(?!\\s*$).+",
    "message" : {
        "pattern" : "repository cant be empty"
    }
}
```
If the repository is empty in the Kubernetes.toml, `repository=""` 

It'll give the following error on the console while compiling. 
ERROR [Kubernetes.toml:(4:12,4:14)] repository cant be empty

pubudu91 commented on misc/toml-parser/src/main/java/io/ballerina/toml/validator/SchemaValidator.java:
```diff
@@ -88,26 +89,43 @@ public void visit(TomlTableNode tomlTableNode) {
         }
     }
 
+    private String getRequiredErrorMessage(String field) {
+        Map<String, String> message = this.schema.message();
+        String typeCustomMessage = message.get(SchemaDeserializer.REQUIRED);
+        if (typeCustomMessage == null) {
+            return "Missing required field \"" + field + "\"";
+        }
+        return typeCustomMessage.replace(PROPERTY_HOLDER, field);
+    }
+
+    private String getUnexpectedPropertyErrorMessage(String key) {
+        Map<String, String> message = this.schema.message();
+        String typeCustomMessage = message.get(SchemaDeserializer.ADDITIONAL_PROPERTIES);
+        if (typeCustomMessage == null) {
+            return "Unexpected Property \"" + key + "\"";
+        }
+        return typeCustomMessage.replace(PROPERTY_HOLDER, key);
+    }
```
These errors get printed in the format you mentioned right? I think we need to revisit the format and phrasing of these sort of error messages to align with the other compiler/parser error messages. e.g., start with lower case letters, using single quotes to quote keys/fields etc.

pubudu91 replied:
@MaryamZi

pramodya1994 replied:
@pubudu91 @MaryamZi  I'm currently integrating this toml parser with schema validation to replace the validations we do in the `ManifestBuilder` class. I will mention you in the PR, once I'm done with it.

xlight05 replied:
@pubudu91 Yes. removed caps and double quotes from all the default error messages and test cases in the recent commit.

pubudu91 commented on misc/toml-parser/src/main/java/io/ballerina/toml/validator/SchemaValidator.java:
```diff
@@ -88,26 +89,43 @@ public void visit(TomlTableNode tomlTableNode) {
+        Map<String, String> message = this.schema.message();
+        String typeCustomMessage = message.get(SchemaDeserializer.ADDITIONAL_PROPERTIES);
+        if (typeCustomMessage == null) {
+            return "unexpected property '" + key + "'";
```
Shall we rephrase this error message so that from the error message itself it's evident that the key/val pair is invalid because it's not in the schema? Maybe something like,
```
key 'foo' not supported in schema 'MySchema'
```

pubudu91 commented on misc/toml-parser/src/main/java/io/ballerina/toml/validator/SchemaValidator.java:
```diff
@@ -130,13 +148,31 @@ public void visit(TomlValueNode tomlValue) {
+        Map<String, String> message = this.schema.message();
+        String typeCustomMessage = message.get(SchemaDeserializer.TYPE);
+        if (typeCustomMessage == null) {
+            return String.format("key '%s' expects %s . found %s", this.key, schema.type(), found);
```
How about something like,
```
incompatible type for key 'foo': expected 'T1', found 'T2'
```

pramodya1994 replied:
@pubudu538 ATM T1 values are all capital. eg: `ARRAY`. Shall we make it like `array`?

pubudu91 commented on misc/toml-parser/src/main/java/io/ballerina/toml/validator/SchemaValidator.java:
```diff
@@ -130,13 +148,31 @@ public void visit(TomlValueNode tomlValue) {
+        Map<String, String> message = this.schema.message();
+        String typeCustomMessage = message.get(SchemaDeserializer.PATTERN);
+        if (typeCustomMessage == null) {
+            return String.format("key '%s' value does not match the regex provided in schema %s", this.key,
```
How about,
```
value for key 'foo' expected to match the regex: <<pattern>>
```

pubudu91 commented on misc/toml-parser/src/main/java/io/ballerina/toml/validator/SchemaValidator.java:
```diff
@@ -190,38 +221,53 @@ public void visit(TomlArrayValueNode tomlArrayValueNode) {
+        Map<String, String> message = this.schema.message();
+        String maxCustomMessage = message.get(SchemaDeserializer.MAXIMUM);
+        if (maxCustomMessage == null) {
+            return String.format("key '%s' value can't be higher than %f", this.key,
+                    max);
+        }
+        return maxCustomMessage;
+    }
+
+    private String getMinValueDeceedErrorMessage(Double min) {
+        Map<String, String> message = this.schema.message();
+        String minCustomMessage = message.get(SchemaDeserializer.MINIMUM);
+        if (minCustomMessage == null) {
+            return String.format("key '%s' value can't be lower than %f", this.key,
+                    min);
```
`value for key 'foo'` instead of `key 'foo' value` for these?
